torch/csrc/utils/future.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+
+namespace torch {
+
+namespace utils {
+
+// FutuerError inherits from std::exception, it can return const char* or
+// std::string error message
+struct TORCH_API FutureError final : public std::exception {
+  FutureError(std::string&& error_msg_) : error_msg(std::move(error_msg_)) {}
+
+  FutureError() = default;
+
+  const char* what() const noexcept override {
+    return error_msg.c_str();
+  }
+
+  std::string errMsg() const {
+    return error_msg;
+  }
+
+  std::string error_msg;
+};
+
+// This class holds a value of type T that will be ready in the future.
+// Most implementation is copied from FutureMessage and
+// c10::ivalue::Future
+template <typename T>
+class TORCH_API Future final {
+ public:
+  using Callback = std::function<void(const T&, bool, const FutureError&)>;
+
+  const T& wait() {
+    std::unique_lock<std::mutex> lock(mutex_);
+    finished_cv_.wait(lock, [this] { return completed_.load(); });
+
+    if (has_error_) {
+      throw error_;
+    }
+    return value_;
+  }
+
+  void markCompleted(T value) {
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      TORCH_CHECK(!completed());
+      completed_ = true;
+      value_ = std::move(value);
+
+      fireCallbacks();
+    }
+    finished_cv_.notify_all();
+  }
+
+  void markCompleted() {
+    markCompleted(T());
+  }
+
+  void markCompleted(FutureError&& error) {
+    std::unique_lock<std::mutex> lock(mutex_);
+    AT_ASSERT(!completed());
+    completed_ = true;
+    has_error_ = true;
+    error_ = std::move(error);
+
+    fireCallbacks();
```

Comment:
lock should be dropped when firing callbacks (this is an existing problem in FutureMessage).
maybe move body of fireCallbacks() here/above to facilitate?

Reply:
This might require more changes to drop the lock on `fireCallbacks()`, otherwise, `addCallback` and `fireCallbacks` can fire concurrently, creating a race on `callbacks_`

Reply:
The races can be avoided because fireCallbacks() clears the callback list (mostly to guarantee that callbacks are called exactly once).

All we need is something like:

```
void makeCompleted(...) {
    std::unique_lock<std::mutex> lock(mutex_);
    ...
    std::vector<Callbacks> cbs;
    cbs.swap(callbacks_);
    lock.unlock();
    for (auto& callback : cb) {
      callback(value_);
    }
}

```
Which will be fine when used in conjunction with the existing addCallback() call, and avoids the deadlock that can occur in the (currently existing FutureMessage) code.

Reply:
will add this

Reply:
@jjlilley this is a good suggestion, and I'm planning to add this in a separated follow up PR to test whether it will cause flakiness or not.